clr/commands.py

```diff
@@ -421,28 +421,55 @@ class NamespaceCache:
 
     This allows subsequent calls to `clr help` or `clr completion_*` to be fast.
     Necessary to work around the fact that many clr command namespace modules
-    import the world and initialize state on import.
+    import the world and initialize state on import. The cache is used only for
+    help and completion calls, when actually calling a clr command its spec is
+    always loaded from the module.
     """
 
     def __init__(self):
+        # Place the cache file in the temp dir with a static file name. Data stored in
+        # this file is easily regeneratable in the event of a system restart, but should
+        # be left behind after the process is complete for subsequent clr calls to find.
         tmpdir = os.environ.get("TMPDIR", "/tmp")
         self.cache_fn = os.path.join(tmpdir, "clr_command_cache")
-        # Clr processes are short lived. We don't close the shelve, but are
-        # careful to sync it after writes.
-        self.cache = shelve.open(self.cache_fn)
+
+        try:
+            # Cache is stored on disk as a shelve, but loaded into memory and then
+            # closed right away so that multiple processes won't conflict trying to
+            # access it.
+            with shelve.open(self.cache_fn) as cache_shelve:
+                self.cache = dict(cache_shelve)
+        except Exception as e:
```

Comment:
Can drop the `as e` if `e` won't be used? (`pylint` will warn about this, but we don't have `pylint` in this repo).

Can we avoid catching the base `Exception` class, though, and instead catch the specific class of errors we expect to see?

Reply:
In this specific case I think catching Exception is actually appropriate. The intention is to ignore _anything_ that happens in these two lines because interrupting an actual clr command from running at the expense of the spec cache not working is never desirable.

The pydoc for shelve does not specify what error can be raised. The error that we saw was a `_gdbm.error`
https://github.com/python/cpython/blob/4d4be477055d0f9e415407bcd193788fae04dc11/Modules/_gdbmmodule.c#L705

If we want to be specific-er OSError is probably the right choice. I can change it to that if you feel strongly.